File: src/main/java/com/teambr/bookshelf/common/blocks/rotation/SixWayRotation.java
package com.teambr.bookshelf.common.blocks.rotation;

import net.minecraft.block.BlockPistonBase;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

/**
 * Bookshelf
 * Created by Paul Davis on 7/27/2015
 */
public class SixWayRotation implements IRotation {
    @Override
    public ForgeDirection convertMetaToDirection(int meta) {
        switch (meta) {
            case 0 :
                return ForgeDirection.DOWN;
            case 1 :
                return ForgeDirection.UP;
            case 2 :
                return ForgeDirection.NORTH;
            case 3 :
                return ForgeDirection.SOUTH;
            case 4 :
                return ForgeDirection.WEST;
            case 5 :
                return ForgeDirection.EAST;
            default :
                return ForgeDirection.UNKNOWN;
        }
    }

    @Override
    public int convertDirectionToMeta(ForgeDirection dir) {
        switch (dir) {
            case NORTH :
                return 2;
            case EAST :
                return 5;
            case SOUTH :
                return 3;
            case WEST :
                return 4;
            case UP :
                return 1;
            case DOWN :
                return 0;
            default :
                return 0;
        }
    }

    @Override
    public int getMetaFromEntity(World world, int x, int y, int z, EntityLivingBase entity, ItemStack itemStack) {
       /**
        * 0: Down
        * 1: Up
        * 2: north
        * 3: south
        * 4: west
        * 5: east
        */
        return BlockPistonBase.determineOrientation(world, x, y, z, entity);
    }
}

File: src/main/scala/com/teambr/bookshelf/common/blocks/SixWayRotation.scala
package com.teambr.bookshelf.common.blocks

import Properties
import net.minecraft.block.properties.IProperty
import net.minecraft.block.state.{BlockStateContainer, IBlockState}
import net.minecraft.block.{Block, BlockPistonBase}
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.{IBlockAccess, World}
import net.minecraftforge.common.property.{ExtendedBlockState, IExtendedBlockState, IUnlistedProperty}

/**
 * This file was created for Bookshelf
 *
 * Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 03, 2015
 */
trait SixWayRotation extends Block with FourWayRotation {

    /**
     * Called when the block is placed, we check which way the player is facing and put our value as the opposite of that
     */
    override def onBlockPlacedBy(world: World, pos : BlockPos, state : IBlockState, placer : EntityLivingBase, stack : ItemStack) : Unit = {
        world.setBlockState(pos, getDefaultState.withProperty(Properties.SIX_WAY, BlockPistonBase.getFacingFromEntity(pos, placer)))
    }

    /**
     * Used to say what our block state is
     */
    override def createBlockState() : BlockStateContainer = {
        val listed = new Array[IProperty[_]](1)
        listed(0) = Properties.SIX_WAY
        val unlisted = new Array[IUnlistedProperty[_]](0)
        new ExtendedBlockState(this, listed, unlisted)
    }

    override def getExtendedState(state : IBlockState, world : IBlockAccess, pos : BlockPos) : IBlockState = {
        state match {
            case returnValue : IExtendedBlockState =>
                returnValue.withProperty(Properties.SIX_WAY, world.getBlockState(pos).getValue(Properties.SIX_WAY).asInstanceOf[EnumFacing])
                returnValue
            case _ =>state
        }
    }

    /**
     * Used to convert the meta to state
 *
     * @param meta The meta
     * @return
     */
    override def getStateFromMeta(meta : Int) : IBlockState = getDefaultState.withProperty(Properties.SIX_WAY, EnumFacing.getFront(meta))

    /**
     * Called to convert state from meta
 *
     * @param state The state
     * @return
     */
    override def getMetaFromState(state : IBlockState) = state.getValue(Properties.SIX_WAY).getIndex
}